### PR TITLE
[DTensor] (166598) Fix matmul operations with _StridedShard placement 

### LINKED ISSUE #166598: Matmul sharding rule incorrectly allows for contraction dim with inconsistent shard order
### 🐛 Describe the bug

These two matmuls should be the same but they are not:

```python
import torch
from torch.distributed.tensor import (
    DTensor,
    DeviceMesh,
    distribute_tensor,
    init_device_mesh,
    Partial,
    Replicate,
    Shard,
)
from torch.distributed.tensor.placement_types import _StridedShard
from torch.distributed._local_tensor import (
    local_tensor_mode,
    LocalTensor,
    LocalTensorMode,
)

S = Shard
_SS = _StridedShard

def product(it):
    x = 1
    for i in it:
        x *= i
    return x

def arange_nd(*sizes):
    if len(sizes) == 1 and isinstance(sizes[0], (list, tuple)):
        sizes = sizes[0]
    return torch.arange(product(sizes)).float().view(sizes)

mesh = init_device_mesh("cuda", [2, 2])
A = distribute_tensor(arange_nd(4, 4), mesh, [S(1), S(1)])
B = distribute_tensor(arange_nd(4, 4), mesh, [S(0), S(0)])
print((A @ B).full_tensor())

B2 = distribute_tensor(arange_nd(4, 4), mesh, [_SS(0, split_factor=2), S(0)])
print((A @ B2).full_tensor())
```

Output:

```
tensor([[ 56.,  62.,  68.,  74.],
        [152., 174., 196., 218.],
        [248., 286., 324., 362.],
        [344., 398., 452., 506.]], device='cuda:0')
tensor([[ 56.,  62.,  68.,  74.],
        [152., 174., 196., 218.],
        [248., 286., 324., 362.],
        [344., 398., 452., 506.]], device='cuda:1')
tensor([[ 56.,  62.,  68.,  74.],
        [152., 174., 196., 218.],
        [248., 286., 324., 362.],
        [344., 398., 452., 506.]], device='cuda:2')
tensor([[ 56.,  62.,  68.,  74.],
        [152., 174., 196., 218.],
        [248., 286., 324., 362.],
        [344., 398., 452., 506.]], device='cuda:3')
tensor([[ 52.,  58.,  64.,  70.],
        [148., 170., 192., 214.],
        [244., 282., 320., 358.],
        [340., 394., 448., 502.]], device='cuda:1')tensor([[ 52.,  58.,  64.,  70.],
        [148., 170., 192., 214.],
        [244., 282., 320., 358.],
        [340., 394., 448., 502.]], device='cuda:2')

tensor([[ 52.,  58.,  64.,  70.],
        [148., 170., 192., 214.],
        [244., 282., 320., 358.],
        [340., 394., 448., 502.]], device='cuda:3')
tensor([[ 52.,  58.,  64.,  70.],
        [148., 170., 192., 214.],
        [244., 282., 320., 358.],
        [340., 394., 448., 502.]], device='cuda:0')
```

### Versions

main

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @tianyu-l @XilunWu @SherlockNoMad

### PR DESCRIPTION
**Pull request by stmcgovern**
Fixes  #166598

  Fixes incorrect results when using _StridedShard placement in matrix
  operations (mm, bmm, addmm, baddbmm). _StridedShard uses interleaved
  sharding semantics but was incorrectly treated as regular Shard,
  causing wrong outputs.

  Solution: Detect exotic placements and generate decomposition strategies
  (redistribute → compute) instead of direct einsum strategies. Populates
  redistribute_cost for cost model compatibility.

  POC redistributes to Replicate (conservative). Future optimization:
  redistribute to Shard targets for better performance.

  Test Plan: Added 3 new tests verifying correctness for mm, bmm, addmm
  with _StridedShard on 2D device mesh.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### CODE REVIEW

**Comment by stmcgovern** [comment]
Thanks for the feedback @ezyang. I'm always trying to improve my workflow. 

 In generating these changes, I wanted to 1) separate out strided shards whose layout was not designed to work with einsum strategies (cause of reported bug) and 2) generate a decomposition strategy that assigns an explicit cost for the redistribution. The only decomposition redistribution strategy implemented is to replicate (this is conservative and there is room for better strategies). 

 Maybe 1) is not desirable since `shard_order` will solve this when it is implemented.  

This is somewhat related to the idea here https://github.com/pytorch/pytorch/pull/164995#discussion_r2478840268 Namely, to have both direct einsum and decomposition (not just as a fallback) strategies that have an explicit cost and can be evaluated in the placement optimization problem. In the case of this PR, only a decomposition strategy (conservative one) is generated (since the exotic placements don't work with direct einsum). 



 ---
  Pre-existing Algorithm (main branch)

  Matrix Operation Strategy Generation

  Single-path approach: All placements treated uniformly

  INPUT: op_schema (contains input tensors A, B with placements)
         einsum_equation (e.g., "mk,kn->mn" for mm)

  ALGORITHM:
  1. Generate all valid einsum strategies for the equation
     → Calls gen_einsum_strategies(equation, mesh)
     → Returns all mathematically valid sharding combinations

  2. Filter strategies by shardability
     FOR EACH strategy:
       - Check if input tensors can be sharded per strategy specs
       - IF shardable:
           * Calculate redistribution costs (current → target placements)
           * Add strategy to valid list
       - ELSE: discard strategy

  3. Return OpStrategy with filtered list

  OUTPUT: OpStrategy containing valid strategies with costs

  Key characteristic: Assumes all Shard-like placements have compatible semantics. Bugs with _StridedShard because it passes is_shard() checks but has incompatible distribution.

  ---
  Current Branch Algorithm (166598-strided-shard-matmul)

  Matrix Operation Strategy Generation with Exotic Placement Handling

  Dual-path approach: Exotic placements use decomposition

  INPUT: op_schema (contains input tensors A, B with placements)
         einsum_equation (e.g., "mk,kn->mn" for mm)
         direct_strategy_fn (original _mm_like_strategy)

  ALGORITHM:
  1. Extract tensor input strategies from op_schema

  2. Detect exotic placements
     FOR EACH input_strategy:
       FOR EACH placement in strategy:
         IF isinstance(placement, _StridedShard):
           → has_exotic = True

  3. Branch on exotic detection:

     PATH A (no exotic placements):
     ├─ Call direct_strategy_fn(equation, mesh, op_schema)
     └─ Return original algorithm result (as above)

     PATH B (exotic placements detected):
     ├─ Define safe target: Replicate on all mesh dimensions
     │
     ├─ Generate decomposition strategies:
     │  FOR EACH input:
     │    - Create target_spec = DTensorSpec(mesh, Replicate-all)
     │    - Calculate redistribution costs (current → Replicate)
     │
     ├─ Compute output spec after redistribution:
     │  - Query gen_einsum_strategies for Replicate inputs
     │  - Find matching strategy
     │  - Extract output_spec
     │
     └─ Create OpSpec:
        * input_specs = [Replicate, Replicate, ...]
        * output_spec = computed from einsum
        * redistribute_cost = calculated costs

  4. Return OpStrategy with strategies

  OUTPUT: OpStrategy (direct strategies OR decomposition strategies)

  Key differences from main:
  - Detection phase: Explicitly checks for _StridedShard
  - Decomposition path: Forces redistribution to Replicate before computation
  - Two-phase execution: Redistribute → Compute (instead of direct computation)
  - Conservative: Prioritizes correctness over performance (Replicate has communication overhead)

  ---
  Side-by-Side Comparison

  | Aspect                         | Main (Pre-existing)        | Current Branch                    |
  |--------------------------------|----------------------------|-----------------------------------|
  | Decision point                 | None (uniform path)        | Exotic placement detection        |
  | Paths                          | 1 (direct einsum)          | 2 (direct OR decomposition)       |
  | Exotic handling                | ❌ Assumes compatible       | ✅ Explicit redistribution         |
  | Strategy type                  | Direct computation         | Direct OR redistribute→compute    |
  | Correctness with _StridedShard | ❌ Wrong results            | ✅ Correct results                 |
  | Performance                    | Optimal for standard Shard | Suboptimal for exotic (Replicate) |
  | Lines of code                  | ~30                        | ~130 (adds ~100 for exotic path)  |

  ---
  Algorithmic Complexity

  Pre-existing:
  - Time: O(S) where S = number of einsum strategies
  - Space: O(S)

  Current branch:
  - Best case (no exotic): O(S) — same as pre-existing
  - Worst case (exotic): O(S + I×M) where I = inputs, M = mesh dimensions
    - Detection: O(I×M)
    - Decomposition generation: O(I)
    - Output spec computation: O(S)

  Runtime communication cost:
  - Pre-existing with exotic: Wrong results (N/A)
  - Current branch with exotic: All-gather = O(tensor_size × world_size) per exotic input
